stock_intercompany/models/stock_picking.py

```diff
@@ -1,49 +1,142 @@
+        string="Intercompany picking that created this counterpart picking",
+        copy=False,
+    )
+    intercompany_child_ids = fields.One2many(
+        comodel_name="stock.picking",
+        inverse_name="intercompany_parent_id",
+        string="Intercompany pickings created by this counterpart picking",
+    )
+
+    def _create_counterpart_picking(self, mode):
+        """
+        Create a counterpart picking for the given picking in the given mode:
+        Mode in: Create incoming from outgoing picking
+        Mode out: Create outgoing from incoming picking
+        """
+
+        self.ensure_one()
+        assert mode in ["in", "out"], "Invalid mode: %s" % mode
+
+        # Skip if already has a parent to avoid looping
+        # Also skip if already has children to avoid creating a duplicate
+        if self.intercompany_parent_id or self.intercompany_child_ids:
+            # We might want to sync if there's already a child
+            return
+
+        location = self.location_id if mode == "out" else self.location_dest_id
+        usage = "customer" if mode == "in" else "supplier"
+        counterusage = "supplier" if mode == "in" else "customer"
+
+        if location.usage != usage:
+            return
+
+        company = (
+            self.env["res.company"]
+            .sudo()
+            .search([("partner_id", "=", self.partner_id.id)], limit=1)
+        )
+        # Only create a picking if the recipient is a different company
+        if not company or company == self.company_id:
+            return
+
+        # Restrict to the configured mode on the target company
+        if company.intercompany_picking_creation_mode not in [mode, "both"]:
+            return
+
+        if not self._check_intercompany_company(company, mode):
+            return
+
+        intercompany_type = getattr(company, "intercompany_%s_type_id" % mode)
+
+        warehouse = intercompany_type.warehouse_id or (
+            self.env["stock.warehouse"]
+            .sudo()
+            .search([("company_id", "=", company.id)], limit=1)
+        )
+
+        # Defalut intercompany type to warehouse in/out type
+        intercompany_type = intercompany_type or getattr(warehouse, "%s_type_id" % mode)
+
+        new_picking_vals = self._prepare_counterpart_picking_vals(
+            company, intercompany_type, warehouse, counterusage
+        )
+        picking = self.env["stock.picking"].sudo().create(new_picking_vals)
+        picking.action_confirm()
+        return picking
+
+    def _check_intercompany_company(self, company, mode):
+        """
+        Hook to check if the given company is valid for the given mode.
+        """
+        return True
+
+    def _prepare_counterpart_picking_vals(
+        self, company, intercompany_type, warehouse, counterusage
+    ):
+        """Prepare the values to create the counterpart picking"""
+        vals = {
+            "partner_id": self.env.user.company_id.partner_id.id,
```

Comment:
Do not take the company on the user. It's the company of the picking
```suggestion
            "partner_id": self.company_id.partner_id.id,
```

Reply:
This was in the original module : https://github.com/OCA/multi-company/blob/14.0/stock_intercompany/models/stock_picking.py#L22 

Should I change it anyway?

Reply:
yes it's a bug.
Since v13, the company of the user is not changed when you change the company in the header "company selector". A user belong to one company whatever the company where is logged.
So when you are in company A and create an "in" in company B. The out will be delivered to company A.
So you can fix it, it was a bug

A user always have one company (that do not depend on the seelect